app/precompiled.py

```diff
@@ -512,9 +512,14 @@ def _get_out_of_bounds_pages(src_pdf_bytes):
 
 
 def escape_special_characters_for_regex(string):
-    special_characters = ["[", "^", "$", ".", "|", "?", "*", "+", "(", ")"]
+    # those characters perform functions in regex expressions and have to be escaped:
+    special_characters = ["[", "{", "^", "$", ".", "|", "?", "*", "+", "(", ")"]
     for character in special_characters:
         string = string.replace(character, r"\{}".format(character))
+
+    bad_slashes = re.findall(r"\\[a-mo-rt-zA-Z]", string)
+    for s in list(set(bad_slashes)):
+        string = string.replace(s, "\\" + s)
```

Comment:
i'm a bit nervous about this code. it looks pretty brittle. if someone passes in the string `\n` as part of a larger text , what does that read from the pdf? the string we get back from the pdf should be `b'\\n'` right?

what happens if someone passes in a pdf with a tab character (escaped that is `\t`) in the address? does that get picked up and changed into `\\t`?

i think we might want to rethink how this code works from the first principles and see

* what the string we get out of the extract address block function looks like
* what the regex we need for matching looks like

and then we can work out how best to transform that string to make it into that regex

Reply:
Hmmm, good point about the tabs. Maybe I could add "t" to omitted letters then, likes "s" and "n"? Right now our code throws an exception if there is "\P" or like in the address, so I think fixing that is more pressing than general improvements (which I'd love to do, but they will probably take longer and on my own I don't have ideas on how to do them).

Reply:
```
if someone passes in the string \n as part of a larger text , what does that read from the pdf? the string we get back from the pdf should be b'\\n' right?
```
This PR would not change behaviour around that case.

Reply:
my concern is I think that if someone passes through `\\t` that _isn't_ a tab character. That's a slash character `\` followed by a `t` character.

python only shows you `\t` as a convenience when you print strings. If there's a `\P` in our address, that's two characters, and I think we need to search for `\\P`.

if we just include `\` in the special characters does that work? 🤔 

i dont know this is giving me a headache thinking about it 😅

Reply:
> if we just include \ in the special characters does that work? 🤔

That would stop it from crashing, but it would not redact that address, because our regex would not be the same as the address in the document.

Hmm I still have the letter that broke it last time saved (I used it for testing that check), so if you have time next week we could go on a call and do some testing together, and maybe even come up with something better?

That is if it won't be another amazingly busy week ;)

Reply:
> my concern is I think that if someone passes through \\t that isn't a tab character. That's a slash character \ followed by a t character.

Ahh I see what you mean, you are right, we should definitely add a test case for what if the backslash is already escaped properly :o

Reply:
yeah. if someone writes the text `bucki\ngham palace` in to their text editor, that'll become `bucki\\ngham` under the hood, and we should be preserving that. But the check will skip over it and the regex will end up being `bucki\s*gham` i think, which wouldn't match. aaaargh this is so confusing!

Reply:
Yeah this one is really tough, but at least it's not something introduced by this PR. I gather if we can make it better in one sitting, lets, if we can't, then just lets make sure nothing becomes worse, the exception gets fixed, deploy what we have, and then make a bigger overhaul plan.